caput/mpiarray.py

```diff
@@ -420,44 +420,108 @@ def redistribute(self, axis):
+                                              comm=self.comm)
+            pac, sac, eac = mpiutil.split_all(arr.shape[axis], comm=self.comm)
 
-        # Perform a local transpose on the array to get the axes in the correct order
-        trans_arr = self.view(np.ndarray).transpose(axlist_f).copy()
+            new_shape = np.asarray(self.global_shape)
+            new_shape[axis] = pc
 
-        # Perform the global transpose
-        tmp_gshape = (self.global_shape[self.axis],) + trans_arr.shape[1:]
-        trans_arr = mpiutil.transpose_blocks(trans_arr, tmp_gshape, comm=self.comm)
+            requests_send = []
+            requests_recv = []
 
-        axlist_b = list(range(len(self.shape)))
-        axlist_b.pop(0)
-        last = axlist_b.pop(-1)
-        if self.axis < axis:  # This has to awkwardly depend on the order of the axes
-            axlist_b.insert(self.axis, 0)
-            axlist_b.insert(axis, last)
-        else:
-            axlist_b.insert(axis, last)
-            axlist_b.insert(self.axis, 0)
+            trans_arr = np.empty(new_shape, dtype=arr.dtype)
+            mpitype = mpiutil.typemap(arr.dtype)
+            buffers = list()
+
+            # Cut out the right blocks of the local array to send around
+            blocks = np.array_split(arr, np.insert(eac, 0, sac[0]), axis)[1:]
+
+            # Iterate over all processes row wise
+            for ir in range(self.comm.size):
+
+                # Iterate over all processes column wise
+                for ic in range(self.comm.size):
+
+                    # Construct a unique tag
+                    tag = ir * self.comm.size + ic
+
+                    # Send and receive the messages as non-blocking passes
+                    if self.comm.rank == ir:
+                        # Send the message
+                        request = self.comm.Isend([blocks[ic].flatten(),
+                                                   mpitype], dest=ic, tag=tag)
+
+                        requests_send.append([ir, ic, request])
+
+                    if self.comm.rank == ic:
+                        buffer_shape = np.asarray(new_shape)
+                        buffer_shape[axis] = eac[ic] - sac[ic]
+                        buffer_shape[self.axis] = ear[ir] - sar[ir]
+                        buffers.append(np.ndarray(buffer_shape, dtype=arr.dtype))
+
+                        request = self.comm.Irecv([buffers[ir], mpitype],
+                                             source=ir, tag=tag)
+                        requests_recv.append([ir, ic, request])
+
+            # Wait for all processes to have started their messages
+            self.comm.Barrier()
+
+            # For each node iterate over all sends and wait until completion
+            for ir, ic, request in requests_send:
+
+                stat = mpiutil.MPI.Status()
+
+                request.Wait(status=stat)
+
+                if stat.error != mpiutil.MPI.SUCCESS:
+                    print
```

Comment:
In Python 3 `print` is now a function (and in python 2 with the `__future__` imports. Why this doesn't break immediately with a `SyntaxError` is quite fun (it's the the fact that the "argument" is on the next line that means it can pass).

Reply:
It does.